File: package/src/components/element_editor.js
import AlchemyElementHeader from "./element/header"
import AlchemyElementToolbar from "./element/toolbar"
import AlchemyElementFooter from "./element/footer"

export default {
  props: {
    element: { type: Object, required: true }
  },

  template: `
    <div :id="elementId" :data-element-id="element.id" :data-element-name="element.name" :class="cssClasses">
      <alchemy-element-header :element="element"></alchemy-element-header>
      <template v-if="!element.folded">
        <alchemy-element-toolbar :element="element"></alchemy-element-toolbar>
        <div class="element-content"></div>
        <alchemy-element-footer :element="element"></alchemy-element-footer>
      </template>
    </div>
  `,

  components: {
    AlchemyElementHeader,
    AlchemyElementToolbar,
    AlchemyElementFooter
  },

  data() {
    const element = this.element
    return {
      elementId: `element_${element.id}`
    }
  },

  computed: {
    cssClasses() {
      let classes = ["element-editor"]

      classes.push(
        this.element.contents.length ? "with-contents" : "without-contents"
      )
      classes.push(
        this.element.nestable_elements.length ? "nestable" : "not-nestable"
      )
      classes.push(this.element.taggable ? "taggable" : "not-taggable")
      classes.push(this.element.folded ? "folded" : "expanded")
      if (this.element.compact) classes.push("compact")
      if (this.element.deprecated) classes.push("deprecated")
      classes.push(this.element.fixed ? "fixed" : "not-fixed")
      classes.push(this.element.public ? "visible" : "hidden")

      return classes.join(" ")
    }
  }
}

File: app/decorators/alchemy/element_editor.rb
# frozen_string_literal: true

module Alchemy
  class ElementEditor < SimpleDelegator
    alias_method :element, :__getobj__

    def to_partial_path
      "alchemy/admin/elements/element"
    end

    # Returns ingredient instances for defined ingredients
    #
    # Creates ingredient on demand if the ingredient is not yet present on the element
    #
    # @return Array<Alchemy::Ingredient>
    def ingredients
      ingredient_definitions.map do |ingredient|
        find_or_create_ingredient(ingredient)
      end
    end

    # Are any ingredients defined?
    # @return [Boolean]
    def has_ingredients_defined?
      ingredient_definitions.any?
    end

    # Returns the translated ingredient group for displaying in admin editor group headings
    #
    # Translate it in your locale yml file:
    #
    #   alchemy:
    #     element_groups:
    #       foo: Bar
    #
    # Optionally you can scope your ingredient role to an element:
    #
    #   alchemy:
    #     element_groups:
    #       article:
    #         foo: Baz
    #
    def translated_group(group)
      Alchemy.t(
        group,
        scope: "element_groups.#{element.name}",
        default: Alchemy.t("element_groups.#{group}", default: group.humanize)
      )
    end

    # CSS classes for the element editor partial.
    def css_classes
      [
        "element-editor",
        ingredient_definitions.present? ? "with-ingredients" : "without-ingredients",
        nestable_elements.any? ? "nestable" : "not-nestable",
        taggable? ? "taggable" : "not-taggable",
        folded ? "folded" : "expanded",
        compact? ? "compact" : nil,
        deprecated? ? "deprecated" : nil,
        fixed? ? "is-fixed" : "not-fixed",
        public? ? nil : "element-hidden"
      ]
    end

    # Tells us, if we should show the element footer and form inputs.
    def editable?
      ingredient_definitions.any? || taggable?
    end

    # Returns ingredients that are not part of any group
    def ungrouped_ingredients
      ingredients.reject { _1.definition.group }
    end

    # Returns ingredients grouped by their group name
    #
    # @return [Hash<String, Array<Alchemy::Ingredient>>]
    def grouped_ingredients
      ingredients.select { _1.definition.group }.group_by { _1.definition.group }
    end

    # Fixes Rails partial renderer calling to_model on the object
    # which reveals the delegated element instead of this decorator.
    def respond_to?(*args, **kwargs)
      method_name = args.first
      return false if method_name == :to_model

      super
    end

    private

    def find_or_create_ingredient(definition)
      element.ingredients.detect { _1.role == definition.role } ||
        element.ingredients.create!(
          role: definition.role,
          type: Alchemy::Ingredient.normalize_type(definition.type)
        )
    end
  end
end
